### ISSUE TITLE
Cargo-semver reports making a trait private that was not public before (`trait_now_doc_hidden`)

### ISSUE THREAD

**Issue by weiznich** [obi1kenobi/cargo-semver-checks#1267]
### Which lint or lints are the issue

trait_now_doc_hidden

### Known issues that might be causing this

- [ ] Is the flagged item defined in another crate, or a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))

### Steps to reproduce the bug with the above code

```
git clone https://github.com/diesel-rs/diesel
cd diesel
git checkout bdb9b254
"cargo" "semver-checks" "-p" "diesel" "--only-explicit-features" "--baseline-version" "2.2.0" "--release-type" "minor" "--features" "sqlite" "--features" "postgres" "--features" "mysql" "--features" "extras"
```

### Actual Behaviour

Description:
A pub trait is now #[doc(hidden)], removing it from the crate's public API.
        ref: https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#hidden
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_now_doc_hidden.ron

Failed in:
  trait IntoArrayExpression in file /home/weiznich/Documents/rust/diesel/diesel/src/pg/expression/array.rs:74

### Expected Behaviour

Cargo semver-check does not report this trait at all as none of the 2.2.x documentation (neither the official one [here](), nor that one locally build from the 2.2.0 release with exactly the same flags) contain references to a `IntoArrayExpression` trait, so this was not public to begin with.

### Verbose Lint Output

"cargo" "semver-checks" "-p" "diesel" "--only-explicit-features" "--baseline-version" "2.2.0" "--release-type" "minor" "--features" "sqlite" "--features" "postgres" "--features" "mysql" "--features" "extras" --verbose
    Updating crates.io index
     Locking 144 packages to latest compatible versions
    Building diesel v2.2.4 (current)
   Compiling dsl_auto_type v0.1.0 (/home/weiznich/Documents/rust/diesel/dsl_auto_type)
   Compiling diesel_derives v2.2.0 (/home/weiznich/Documents/rust/diesel/diesel_derives)
 Documenting diesel v2.2.4 (/home/weiznich/Documents/rust/diesel/diesel)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 4.07s
       Built [   4.142s] (current)
     Parsing diesel v2.2.4 (current)
      Parsed [   0.163s] (current)
     Parsing diesel v2.2.0 (baseline, cached)
      Parsed [   0.122s] (baseline)
    Checking diesel v2.2.0 -> v2.2.4 (assume minor change)
    Starting 128 checks, 36 unnecessary on 20 threads
        PASS [   0.004s]       major        attribute_proc_macro_missing
        PASS [   0.041s]       major        auto_trait_impl_removed
        PASS [   0.008s]       major        constructible_struct_adds_field
        PASS [   0.010s]       major        constructible_struct_adds_private_field
        PASS [   0.006s]       major        constructible_struct_changed_type
        PASS [   0.004s]       major        declarative_macro_missing
        PASS [   0.004s]       major        derive_helper_attr_removed
        PASS [   0.004s]       major        derive_proc_macro_missing
        PASS [   0.059s]       major        derive_trait_impl_removed
        PASS [   0.004s]       major        enum_discriminants_undefined_non_exhaustive_variant
        PASS [   0.007s]       major        enum_discriminants_undefined_non_unit_variant
        PASS [   0.017s]       major        enum_marked_non_exhaustive
        PASS [   0.004s]       major        enum_missing
        PASS [   0.006s]       major        enum_no_repr_variant_discriminant_changed
        PASS [   0.004s]       major        enum_non_exhaustive_tuple_variant_changed_kind
        PASS [   0.004s]       major        enum_now_doc_hidden
        PASS [   0.004s]       major        enum_repr_int_changed
        PASS [   0.004s]       major        enum_repr_int_removed
        PASS [   0.004s]       major        enum_repr_transparent_removed
        PASS [   0.004s]       major        enum_repr_variant_discriminant_changed
        PASS [   0.004s]       major        enum_struct_variant_changed_kind
        PASS [   0.008s]       major        enum_struct_variant_field_added
        PASS [   0.004s]       major        enum_struct_variant_field_missing
        PASS [   0.004s]       major        enum_struct_variant_field_now_doc_hidden
        PASS [   0.008s]       major        enum_tuple_variant_changed_kind
        PASS [   0.012s]       major        enum_tuple_variant_field_added
        PASS [   0.013s]       major        enum_tuple_variant_field_missing
        PASS [   0.004s]       major        enum_tuple_variant_field_now_doc_hidden
        PASS [   0.005s]       major        enum_unit_variant_changed_kind
        PASS [   0.008s]       major        enum_variant_added
        PASS [   0.005s]       major        enum_variant_marked_non_exhaustive
        PASS [   0.005s]       major        enum_variant_missing
        PASS [   0.008s]       major        exported_function_changed_abi
        PASS [   0.001s]       major        feature_missing
        PASS [   0.000s]       major        feature_not_enabled_by_default
        PASS [   0.007s]       major        function_abi_no_longer_unwind
        PASS [   0.008s]       major        function_abi_now_unwind
        PASS [   0.004s]       major        function_changed_abi
        PASS [   0.004s]       major        function_const_removed
        PASS [   0.004s]       major        function_export_name_changed
        PASS [   0.003s]       major        function_like_proc_macro_missing
        PASS [   0.004s]       major        function_missing
        PASS [   0.004s]       major        function_now_doc_hidden
        PASS [   0.004s]       major        function_parameter_count_changed
        PASS [   0.004s]       major        function_requires_different_const_generic_params
        PASS [   0.004s]       major        function_requires_different_generic_type_params
        PASS [   0.003s]       major        function_unsafe_added
        PASS [   0.004s]       major        inherent_associated_const_now_doc_hidden
        PASS [   0.005s]       major        inherent_associated_pub_const_missing
        PASS [   0.005s]       major        inherent_method_const_removed
        FAIL [   0.007s]       major        inherent_method_missing
        PASS [   0.007s]       major        inherent_method_now_doc_hidden
        PASS [   0.006s]       major        inherent_method_unsafe_added
        PASS [   0.003s]       major        macro_no_longer_exported
        PASS [   0.003s]       major        macro_now_doc_hidden
        PASS [   0.014s]       major        method_parameter_count_changed
        PASS [   0.010s]       major        method_requires_different_const_generic_params
        PASS [   0.009s]       major        method_requires_different_generic_type_params
        PASS [   0.004s]       major        module_missing
        PASS [   0.004s]       major        non_exhaustive_struct_changed_type
        PASS [   0.004s]       major        partial_ord_enum_struct_variant_fields_reordered
        PASS [   0.004s]       major        partial_ord_enum_variants_reordered
        PASS [   0.012s]       major        partial_ord_struct_fields_reordered
        PASS [   0.003s]       major        proc_macro_now_doc_hidden
        PASS [   0.003s]       major        pub_module_level_const_missing
        PASS [   0.011s]       major        pub_module_level_const_now_doc_hidden
        PASS [   0.003s]       major        pub_static_missing
        PASS [   0.003s]       major        pub_static_mut_now_immutable
        PASS [   0.005s]       major        pub_static_now_doc_hidden
        PASS [   0.004s]       major        pub_static_now_mutable
        PASS [   0.004s]       major        repr_c_enum_struct_variant_fields_reordered
        PASS [   0.005s]       major        repr_c_plain_struct_fields_reordered
        PASS [   0.004s]       major        repr_c_removed
        PASS [   0.009s]       major        repr_packed_added
        PASS [   0.004s]       major        repr_packed_removed
        PASS [   0.013s]       major        sized_impl_removed
        PASS [   0.003s]       major        static_became_unsafe
        PASS [   0.008s]       major        struct_marked_non_exhaustive
        PASS [   0.006s]       major        struct_missing
        PASS [   0.007s]       major        struct_now_doc_hidden
        PASS [   0.006s]       major        struct_pub_field_missing
        PASS [   0.006s]       major        struct_pub_field_now_doc_hidden
        PASS [   0.004s]       major        struct_repr_transparent_removed
        PASS [   0.007s]       major        struct_with_no_pub_fields_changed_type
        PASS [   0.004s]       major        struct_with_pub_fields_changed_type
        FAIL [   0.008s]       major        trait_added_supertrait
        PASS [   0.005s]       major        trait_allows_fewer_const_generic_params
        PASS [   0.013s]       major        trait_allows_fewer_generic_type_params
        PASS [   0.008s]       major        trait_associated_const_added
        PASS [   0.007s]       major        trait_associated_const_default_removed
        PASS [   0.003s]       major        trait_associated_const_now_doc_hidden
        PASS [   0.008s]       major        trait_associated_type_added
        PASS [   0.009s]       major        trait_associated_type_default_removed
        PASS [   0.005s]       major        trait_associated_type_now_doc_hidden
        PASS [   0.009s]       major        trait_method_added
        PASS [   0.008s]       major        trait_method_default_impl_removed
        PASS [   0.010s]       major        trait_method_missing
        PASS [   0.009s]       major        trait_method_now_doc_hidden
        PASS [   0.014s]       major        trait_method_parameter_count_changed
        PASS [   0.012s]       major        trait_method_requires_different_const_generic_params
        PASS [   0.013s]       major        trait_method_requires_different_generic_type_params
        PASS [   0.008s]       major        trait_method_unsafe_added
        PASS [   0.003s]       major        trait_method_unsafe_removed
        PASS [   0.006s]       major        trait_mismatched_generic_lifetimes
        PASS [   0.006s]       major        trait_missing
        PASS [   0.003s]       major        trait_newly_sealed
        FAIL [   0.008s]       major        trait_no_longer_dyn_compatible
        FAIL [   0.004s]       major        trait_now_doc_hidden
        PASS [   0.004s]       major        trait_removed_associated_constant
        PASS [   0.005s]       major        trait_removed_associated_type
        PASS [   0.006s]       major        trait_removed_supertrait
        PASS [   0.009s]       major        trait_requires_more_const_generic_params
        PASS [   0.009s]       major        trait_requires_more_generic_type_params
        PASS [   0.004s]       major        trait_unsafe_added
        PASS [   0.003s]       major        trait_unsafe_removed
        PASS [   0.003s]       major        tuple_struct_to_plain_struct
        PASS [   0.008s]       major        type_allows_fewer_const_generic_params
        PASS [   0.008s]       major        type_allows_fewer_generic_type_params
        PASS [   0.008s]       major        type_mismatched_generic_lifetimes
        PASS [   0.007s]       major        type_requires_more_const_generic_params
        PASS [   0.009s]       major        type_requires_more_generic_type_params
        PASS [   0.003s]       major        union_field_added_with_all_pub_fields
        PASS [   0.003s]       major        union_field_added_with_non_pub_fields
        PASS [   0.003s]       major        union_field_missing
        PASS [   0.003s]       major        union_missing
        PASS [   0.003s]       major        union_now_doc_hidden
        PASS [   0.003s]       major        union_pub_field_now_doc_hidden
        PASS [   0.004s]       major        unit_struct_changed_kind
     Checked [   0.063s] 128 checks: 124 pass, 4 fail, 0 warn, 36 skip

--- failure inherent_method_missing: pub method removed or renamed ---

Description:
A publicly-visible method or associated fn is no longer available under its prior name. It may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/inherent_method_missing.ron

Failed in:
  SerializedDatabase::new, previously in file /home/weiznich/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/diesel-2.2.0/src/sqlite/connection/serialized_database.rs:16

--- failure trait_added_supertrait: non-sealed trait added new supertraits ---

Description:
A non-sealed trait added one or more supertraits, which breaks downstream implementations of the trait
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#generic-bounds-tighten
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_added_supertrait.ron

Failed in:
  trait diesel::connection::Instrumentation gained Downcast in file /home/weiznich/Documents/rust/diesel/diesel/src/connection/instrumentation.rs:246

--- failure trait_no_longer_dyn_compatible: trait no longer dyn compatible ---

Description:
Trait is no longer dyn compatible, which breaks `dyn Trait` usage.
        ref: https://doc.rust-lang.org/stable/reference/items/traits.html#object-safety
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_no_longer_dyn_compatible.ron

Failed in:
  trait SqlOrd in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/ord.rs:7
  trait SqlType in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/mod.rs:671
  trait Foldable in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/fold.rs:4
  trait SingleValue in file /home/weiznich/Documents/rust/diesel/diesel/src/sql_types/mod.rs:655

--- failure trait_now_doc_hidden: pub trait is now #[doc(hidden)] ---

Description:
A pub trait is now #[doc(hidden)], removing it from the crate's public API.
        ref: https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#hidden
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_now_doc_hidden.ron

Failed in:
  trait IntoArrayExpression in file /home/weiznich/Documents/rust/diesel/diesel/src/pg/expression/array.rs:74

     Summary semver requires new major version: 4 major and 0 minor checks failed
    Finished [  23.639s] diesel

### Generated System Information

System information:
-------------------
#### Software version

cargo-semver-checks 0.41.0 (793f25c)

#### Operating system

- OS: Linux (Fedora Linux 42)
- Kernel: 6.14.5-300.fc42.x86_64

#### Command-line

```bash
/home/weiznich/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.87.0 (99624be96 2025-05-06)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

Cargo build configuration:
--------------------------
[alias]
xtask = ["run", "--package", "xtask", "--"]

**Comment by obi1kenobi** [comment]
Running with `--verbose --verbose` shows that this is a true positive with a bad diagnostic message:

```
--- failure trait_now_doc_hidden: pub trait is now #[doc(hidden)] ---

Description:
A pub trait is now #[doc(hidden)], removing it from the crate's public API.
        ref: https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#hidden
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.41.0/src/lints/trait_now_doc_hidden.ron

Failed in:
  trait IntoArrayExpression in file /tmp/diesel/diesel/src/pg/expression/array.rs:74
        lint rule output values:
    {
      "path": [
        "diesel",
        "expression",
        "AsExpressionList"
      ],
      "span_begin_line": 74,
      "span_end_line": 81,
      "span_filename": "/tmp/diesel/diesel/src/pg/expression/array.rs",
      "trait_name": "IntoArrayExpression"
    }
```

`diesel::expression::AsExpressionList` indeed [used to be public API](https://github.com/diesel-rs/diesel/blob/197f06d8d1c0f66774b1ab24c5db93d28962e07a/diesel/src/expression/mod.rs#L1044-L1058) and [has become `doc(hidden)`](https://github.com/diesel-rs/diesel/blob/611e949f197260720ac53e71863a854fd0cae55f/diesel/src/expression/mod.rs#L1058-L1066). We get the trait name `IntoArrayExpression` because that's the actual declared name of the trait that is now being re-exported as `diesel::expression::AsExpressionList`.

We moved away from showing full paths in diagnostics because they were often too long and not helpful. But in this case, it would have been the right thing to show. Perhaps we can use a heuristic:
- show only the name if that name matches the last component of the path
- otherwise show the path instead.

The diagnostic would then have said:
```
Failed in:
  trait diesel::expression::AsExpressionList in file /tmp/diesel/diesel/src/pg/expression/array.rs:74
```
Regrettably, this would still show the trait's definition span and not the re-export. That's a much more difficult problem to solve, unfortunately.

What do you think?
